### PR TITLE
feat(CollectibleV1): add `safeBatchTransferFrom` capabilities

### LINKED ISSUE #41: Implement batch transfer extension in `BaseToken`
**UPDATE: The description below is outdated. This issue has been update according to https://github.com/status-im/communities-contracts/issues/41#issuecomment-1953630995**

Status communities want to be able to batch transact tokens. Hence there's a request to move away from `CollectibleV1` and instead have an ERC1155 implementation. Essentially it should be a drop-in replacement, meaning that, in addition to being an ERC1155 implementation, it should also have the custom function related to things like `maxSupply` that are currently implemented by `CollectibleV1` and inherently `BaseToken`.

There is a chance we can't use `BaseToken` for this anymore, but this needs to analyzed.

To close this issue we need:

- [x] Smart contract implementation
- [x] Natspec docs where applicable
- [x] Unit tests

### PR DESCRIPTION
**Pull request by 0x-r4bbit**
This is to allow batch transfers of community collectibles as discussed in #41.

Closes #41

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Added natspec comments?
- [x] Ran `forge snapshot`?
- [x] Ran `pnpm lint`?
- [x] Ran `forge test`?
- [x] Ran `pnpm verify`?

### CODE REVIEW

File: contracts/tokens/BaseToken.sol [hunk 1]
```diff
@@ -153,5 +164,32 @@ abstract contract BaseToken is Context, ERC721Enumerable, CommunityOwnable {
         super._beforeTokenTransfer(from, to, firstTokenId, batchSize);
     }
 
+    /**
+     * @dev See {IERC1155-safeBatchTransferFrom}.
+     */
```
Comment (0x-r4bbit) [inline contracts/tokens/BaseToken.sol:169]: This natspec needs to change as our `safeBatchTransfer()` implementation isn't actually `ERC1155` compliant.

File: contracts/tokens/BaseToken.sol [hunk 2]
```diff
@@ -153,5 +164,32 @@ abstract contract BaseToken is Context, ERC721Enumerable, CommunityOwnable {
...
+    )
+        internal
+        virtual
+    {
+        for (uint256 i = 0; i < ids.length; ++i) {
+            uint256 id = ids[i];
+            _safeTransfer(from, to, id, data);
+        }
+        address operator = _msgSender();
+        emit TransferBatch(operator, from, to, ids);
```
Comment (0x-r4bbit) [inline contracts/tokens/BaseToken.sol:191]: I've added this event here for completeness-sake, but want to point out that each individual `safeTransfer()` call will emit its own `Transfer` event.

@gravityblast lemme know if you think this should be removed.

File: test/CollectibleV1.t.sol [hunk 3]
```diff
@@ -116,3 +116,50 @@ contract RemoteBurnTest is CollectibleV1Test {
...
+        ids[0] = 0;
+        ids[1] = 1;
+        ids[2] = 2;
+
+        vm.prank(accounts[0]);
+        collectibleV1.safeBatchTransferFrom(accounts[0], accounts[1], ids, "");
+
+        assertEq(collectibleV1.balanceOf(accounts[0]), 0);
+        assertEq(collectibleV1.balanceOf(accounts[1]), 3);
+    }
```
Comment (0x-r4bbit) [inline test/CollectibleV1.t.sol:164]: Can probably add some more tests here.

File: contracts/tokens/BaseToken.sol [hunk 4]
```diff
@@ -153,5 +164,32 @@ abstract contract BaseToken is Context, ERC721Enumerable, CommunityOwnable {
         super._beforeTokenTransfer(from, to, firstTokenId, batchSize);
     }
 
+    /**
+     * @dev See {IERC1155-safeBatchTransferFrom}.
+     */
+    function safeBatchTransferFrom(address from, address to, uint256[] memory ids, bytes memory data) public virtual {
+        if (from != _msgSender() && !isApprovedForAll(from, _msgSender())) {
+            revert BaseToken_NotAuthorized();
```
Comment (0xb337r007) [inline contracts/tokens/BaseToken.sol:172]: internally  the ERC721 `transferFrom` is called and it's already checking for ownership and approvals, so if we remove this we save some gas

File: contracts/tokens/BaseToken.sol (hunk 2 above)
Comment (0x-r4bbit) [reply contracts/tokens/BaseToken.sol:191]: Okay then I'll remove this for now.

File: contracts/tokens/BaseToken.sol (hunk 4 above)
Comment (0x-r4bbit) [reply contracts/tokens/BaseToken.sol:172]: Actually I've used `_safeTransfer()` which ultimately uses `_transfer()` which does no checks on `msg.sender` or `approvals`. 

Thank you for pointing it out. We either keep this check here and use `_safeTransfer()` or, we remove this check here but then have to use `safeTransferFrom()` instead.

I'm gonna change this to use `safeTransferFrom()` and remove the check here.

Thanks again!

File: contracts/tokens/BaseToken.sol (hunk 2 above)
Comment (0x-r4bbit) [reply contracts/tokens/BaseToken.sol]: Removed the event.

File: test/CollectibleV1.t.sol (hunk 3 above)
Comment (0x-r4bbit) [reply test/CollectibleV1.t.sol:173]: I've updated the `RevertWhen_NotAuthorized` test to check for the approvals case.

File: contracts/tokens/BaseToken.sol (hunk 4 above)
Comment (0x-r4bbit) [reply contracts/tokens/BaseToken.sol]: Updated this as mentioned above.

File: contracts/tokens/BaseToken.sol [hunk 5]
```diff
@@ -153,5 +155,62 @@ abstract contract BaseToken is Context, ERC721Enumerable, CommunityOwnable {
...
+        public
+        virtual
+    {
+        address[] memory receivers = new address[](ids.length);
+        uint256 i = 0;
+        for (; i < ids.length; i++) {
+            receivers[i] = to;
+        }
+        _safeBatchTransferFrom(from, receivers, ids, data);
+    }
```
Comment (0x-r4bbit) [inline contracts/tokens/BaseToken.sol:194]: @gravityblast I've added an overloaded version of this function to take a single `to` address as well, for ergonomics. If you think this is not worth the cost, I'm going to remove it again.

File: contracts/tokens/BaseToken.sol [hunk 6]
```diff
@@ -153,5 +154,62 @@ abstract contract BaseToken is Context, ERC721Enumerable, CommunityOwnable {
...
+        internal
+    {
+        if (receivers.length != ids.length) {
+            revert BaseToken_ReceiversAndIdsMismatch();
+        }
+
+        for (uint256 i = 0; i < ids.length; ++i) {
+            address to = receivers[i];
+            uint256 id = ids[i];
+            safeTransferFrom(from, to, id, data);
```
Comment (0xb337r007) [inline contracts/tokens/BaseToken.sol:211]: we might avoid the 2 variables and just use `array[i]` in the parameters, but I leave it up to you, cause maybe it's not a big change in gas

**Review by gravityblast — APPROVED** [review]

**Pull request merged**